### ISSUE TITLE
Install Rolback

### ISSUE THREAD

**Issue by BjornHoorelbeke** [dorssel/usbipd-win#430]
Hi,

I want to install USBIPD on a Windows Server 2012R2 machine and the install is rolled back everytime I try.

This is what I find in the Eventlog : _"Windows Installer installed the product. Product Name: usbipd-win. Product Version: 2.3.0. Product Language: 1033. Manufacturer: Frans van Dorsselaer. Installation success or error status: 1603."_
I looked for this message but found nothing that helped  me.

I then enabled verbose logging on the msi and in attachment is what I got, but I have no idea where to start looking.
[MSILog.txt](https://github.com/dorssel/usbipd-win/files/9324469/MSILog.txt)

Any idea what I'm doing wrong?

**Comment by dorssel** [comment]
Thanks for the report!

The reason was easy to find from the log: the VirtualBox drivers are signed incorrectly by Oracle, so they don't work anymore on anything pre-Win10. Oracle messed the signing procedure up once again... (not the first time, see https://github.com/dorssel/usbipd-win/issues/47#issuecomment-953359682). This time they did the Win10/Win11 bit right, but now they have broken the Pre-Win10 signatures completely.

The fix took a bit of time, but I think I have repaired it by creating and signing a new driver catalog specifically for pre-Win10. Can you try the installer from #431? It is the build artifact here: https://github.com/dorssel/usbipd-win/actions/runs/2853753959.

NOTE: The *drivers* are signed personally by me, but the *installer* is not signed yet. You may get a warning at the beginning of the installation (until I make a new, signed release) and then you will probably get a popup asking if you trust the drivers (which should be properly signed now).

PS: You did nothing wrong, Oracle did...

**Comment by BjornHoorelbeke** [comment]
The install is working flawless now, thx for the quick response!

Unfortunately I won't be able to use your software, I was hoping to share my usb2serial device to home-assistant running inside hyper-v.
But anyway, the install is fixed now ;)

**Comment by dorssel** [comment]
@BjornHoorelbeke 
Thanks for testing. Are running Linux inside Hyper-V? How far did you get?

**Comment by BjornHoorelbeke** [comment]
I succeeded in sharing my device
![image](https://user-images.githubusercontent.com/37407247/185072769-edada13b-8aac-402f-84cf-b22a7e8f6d2b.png)

The Hyper-V is running "Home Assistant OS" which is Linux based. But I don't know how to use the shared device there

**Comment by dorssel** [comment]
@BjornHoorelbeke 
Ah. You did step 1 (share from Windows). Now you need step 2: `usbip attach` from the Linux side, see https://github.com/dorssel/usbipd-win#connecting-devices.

**Comment by BjornHoorelbeke** [comment]
I managed to install usbip on the Home Assistant OS, but now I get the following errors
![image](https://user-images.githubusercontent.com/37407247/185108440-cbbd90b4-e040-46aa-922e-cd53d2a49152.png)

**Comment by dorssel** [comment]
The first error is benign, it just means you haven't installed `hwdata` (or whatever the package is named for your distro), meaning that it will only display numbers instead of names for devices. You can fix that, but it is not really important.

The second error is what makes things fail.

What is the kernel version (`uname -a`)? And what is the usbip version (`usbip version`)? Usually such errors come from a version mismatch between the kernel and the user mode software. Is there any more info in /var/log/syslog (or whatever is the default error log for this distro)?

You can also try if the (windows) server side provides more error details. See https://github.com/dorssel/usbipd-win/wiki/Troubleshooting.

**Comment by BjornHoorelbeke** [comment]
Thx for your help!

these are the versions 
![image](https://user-images.githubusercontent.com/37407247/185554093-b7218212-bdc8-474a-b52b-c1197680b054.png)

I have enabled logging on the server and this is what I got when doing a usbip list
![image](https://user-images.githubusercontent.com/37407247/185554286-93a01717-1abf-41d5-af8c-82e8b036af7a.png)

Nothing to see in the EventLogs

**Comment by dorssel** [comment]
I am (almost) certain that the version of Linux-side `usbip` is too old.

Can you find out which package `usbip` belongs to on your distro? (package could be itself called `usbip`, but might also be called something like `linux-tools-common`). And what is the version of that package?

One other idea is that it is a firewall issue. However, the connect seems to work fine. It just looks like `usbip` is not interpreting the returned data correctly, probably because it is too old and not using the correct protocol version. But there is a remote possibility that the connect is going through the firewall alright, but the returned data is not... You could test that by using `tcpdump` while doing `usbip list`.

Is there anything in syslog on the Linux side?

**Comment by BjornHoorelbeke** [comment]
This is the version I have installed, which is a recent version : https://pkgs.alpinelinux.org/package/edge/community/x86/linux-tools-usbip
![image](https://user-images.githubusercontent.com/37407247/185561075-987f8bbc-1a4b-4328-9bc4-3b1a03835686.png)

I can't find anything in the logs on the Linux side.

I ran tcpdump, but again I have no idea what to look for.

I understand these are all not your worries, and mostly my ignorance. So if you want to close this, I would completely understand

**Comment by dorssel** [comment]
O ... I don't give up easily. Not until it is either fixed, or we have concluded that it is unfixable. And we're not there yet.

I now understand that your distro is (based on) Alpine Linux. The missing hwdata is in package `hwids-usb`, see https://pkgs.alpinelinux.org/package/edge/main/x86_64/hwids-usb.

First, what command were you running when you got the initial error? `usbip attach` or `usbip list`?
Lets try to list the devices first. After installing `hwids-usb`, what is the output of:

```
usbip list -r <HyperV-host-IP-address>
```

(Your host address is most likely in the 192.168.1.x range.)

**Comment by BjornHoorelbeke** [comment]
Thx!

I never tried the attach because I started with the list

![image](https://user-images.githubusercontent.com/37407247/185622703-09b7fdf4-6cdb-449b-8f0e-33c3d2f4976e.png)

the host ip address is 192.168.1.17 and the one of the VM is 192.168.1.92, when debugging usbipd I see a connection from the VM coming in
![image](https://user-images.githubusercontent.com/37407247/185623346-a2492786-be84-4b16-8f34-dc55fd5b0e19.png)

**Comment by dorssel** [comment]
OK, so the hwdata is present now. On error down.

Now, from another Linux terminal run:
```
sudo tcpdump -i any -l port 3240
```

and then do the `usbip list` again in the original terminal. What is the output of tcpdump?

**Comment by dorssel** [comment]
O wow. I found a (small) bug in the Linux kernel. Not the cause of this error, but the cause of the weird error text "Unknown Op Common Status".

Here is the call that fails (I think): https://elixir.bootlin.com/linux/v5.5.15/source/tools/usb/usbip/src/usbip_list.c#L61

It will then try to display the error text for `status`, which is uninitialized before the call (here: https://elixir.bootlin.com/linux/v5.5.15/source/tools/usb/usbip/src/usbip_list.c#L53).

But the call that fails can return *before* it sets status. Like here: https://elixir.bootlin.com/linux/v5.5.15/source/tools/usb/usbip/src/usbip_network.c#L156, or here: https://elixir.bootlin.com/linux/v5.5.15/source/tools/usb/usbip/src/usbip_network.c#L172. `status` does not get set until here: https://elixir.bootlin.com/linux/v5.5.15/source/tools/usb/usbip/src/usbip_network.c#L196.

In any case, I will report it to the kernel team. And for us: the communication fails very early on. This means either firewall issues (connection gets through, but data is being blocked in one or both directions), or there is a version mismatch:
https://elixir.bootlin.com/linux/v5.5.15/source/tools/usb/usbip/src/usbip_network.c#L177.

TL;DR: This is not our error, but is the cause of the weird error *message* that I haven't seen before.

**Comment by BjornHoorelbeke** [comment]
confused by the above ;)

I ran tcpdump while executing usbip list and this is the output
![image](https://user-images.githubusercontent.com/37407247/185632415-05a5c9da-aa74-46c0-906b-551fae240377.png)
Hope this helps?

**Comment by dorssel** [comment]
Yes, a little. Please try again, but with:

```
sudo tcpdump -i any -l port 3240 -X
```

(extra `-X` at the end)

**Comment by BjornHoorelbeke** [comment]
It's getting more and more chinese ;)

![image](https://user-images.githubusercontent.com/37407247/185637148-1d921122-41a6-49bf-b6bf-f5d362a87fca.png)

**Comment by dorssel** [comment]
That looks promising. Packet breakdown:

1) client connect
2) server acknowledges connection
3) client acknowledges connection (normal 3-way TCP handshake)
4) client requests list, the data `0111 8005 0000 0000` is correct. Correct request, correct version, correct byte order.
5) server resets connection <-- this is weird.

However, since the client actually *does* send a request, and the *first* thing the server does is logging any received request:
```
Logger.Debug($"Received opcode: {opCode}");
```

This means the server never got the request data and just hangs up. This smells like a firewall issue. Do you run any third party firewall on your host (other than the default Microsoft firewall)? What is your host OS version, and is it Windows Server or Windows consumer edition?

**Comment by BjornHoorelbeke** [comment]
This is my server version
![image](https://user-images.githubusercontent.com/37407247/185639882-39b6b15d-30a8-4021-be82-46897a743a4c.png)

I only use the builtin firewall and I see that port 3240 is allowed.

**Comment by dorssel** [comment]
New PR https://github.com/dorssel/usbipd-win/pull/432.
This doesn't actually fix anything, but it improves error reporting in the early communication stage.

Installer is at https://github.com/dorssel/usbipd-win/actions/runs/2890493378.
Since this reports the same development version 2.3.1 as you have now, you will have to uninstall first.

Could you give it a try? Thx.

**Comment by BjornHoorelbeke** [comment]
Thank you!

Is this the feedback you want to see?
![image](https://user-images.githubusercontent.com/37407247/185657840-443dc7b9-7d10-42c9-9bb1-fcad0d23eba0.png)

**Comment by dorssel** [comment]
Yes, this confirms that the server is not receiving any bytes from the client. If it would have received at least 1 byte, it would have logged it.

So:
1) client connects
2) server accepts connection
3) client sends request (confirmed by tcpdump)
4) server *does not* receive request (at all, not a single byte)

This must be caused by some networking issues between the host and the VM. Are you also connecting to the VM using the same IP address? If so, it must be port-related.

Can you try on the VM:
```
nc 192.168.1.17 3240
```
and then type 1 character + ENTER, followed by Ctrl+C. What does `uspipd` log in that case.
Here is what I would get:

```
dbug: Microsoft.Extensions.Hosting.Internal.Host[1]
      Hosting starting
dbug: Usbipd.PcapNg[1000]
      usbipd:PcapNg:Path = ''
dbug: Usbipd.Server[1000]
      2.3.1-error-reporting.1+75.Branch.error-reporting.Sha.8181be0d3446e9926f75b1fe92d851ff8f19c597
info: Microsoft.Hosting.Lifetime[0]
      Application started. Press Ctrl+C to shut down.
info: Microsoft.Hosting.Lifetime[0]
      Hosting environment: Production
info: Microsoft.Hosting.Lifetime[0]
      Content root path: C:\Users\Frans\Source\Repos\usbipd-win\Usbipd\bin\x64\Debug\net6.0-windows8.0\win-x64
dbug: Microsoft.Extensions.Hosting.Internal.Host[2]
      Hosting started
dbug: Usbipd.Server[1000]
      new connection from 192.168.99.11
fail: Usbipd.ConnectedClient[3]
      An exception occurred while communicating with the client:
      System.Net.ProtocolViolationException: client disconnect in the middle of a message, received 2 from 8 bytes
         at Usbipd.Tools.ReadExactlyAsync(Stream stream, Memory`1 buf, CancellationToken cancellationToken) in C:\Users\Frans\Source\Repos\usbipd-win\Usbipd\Tools.cs:line 39
         at Usbipd.ConnectedClient.RecvOpCodeAsync(CancellationToken cancellationToken) in C:\Users\Frans\Source\Repos\usbipd-win\Usbipd\ConnectedClient.cs:line 283
         at Usbipd.ConnectedClient.RunAsync(CancellationToken cancellationToken) in C:\Users\Frans\Source\Repos\usbipd-win\Usbipd\ConnectedClient.cs:line 56
dbug: Usbipd.Server[1000]
      connection closed
```

**Comment by dorssel** [comment]
PS: reported the weird error reporting by the Linux kernel: https://marc.info/?l=linux-usb&m=166092303904227&w=2. Guess I'll have to fix it now too...

**Comment by BjornHoorelbeke** [comment]
Hmm, doesn't seem to give a lot of new info. Looks like the character doesn't end up on the server side
![image](https://user-images.githubusercontent.com/37407247/185679886-74e48bbf-0c69-4629-8221-da198767afd3.png)

**Comment by BjornHoorelbeke** [comment]
I disabled the firewall on the server and it doesn't make a difference

**Comment by dorssel** [comment]
*When* do you see connection closed? Immediately after connect, or only after you press Ctrl+C ?

**Comment by dorssel** [comment]
Another idea, just to check if the server is capable of receiving at all. From the Windows host itself, in PowerShell, run the following commands:
```
$tcpClient = New-Object System.Net.Sockets.TCPClient
$tcpClient.Connect("localhost",3240)
$tcpClient.GetStream().WriteByte(1)
$tcpClient.Close()
```
That should produce the new 'ProtocolViolationException: client disconnect in the middle of a message' error.

**Comment by BjornHoorelbeke** [comment]
> _When_ do you see connection closed? Immediately after connect, or only after you press Ctrl+C ?

The below happens as soon as I execute the "nc" command. So before I even send a character or press Ctrl+C

![image](https://user-images.githubusercontent.com/37407247/185730223-00288863-8bca-4ec6-bf20-a0d7310c8398.png)

**Comment by BjornHoorelbeke** [comment]
> Another idea, just to check if the server is capable of receiving at all. From the Windows host itself, in PowerShell, run the following commands:
> 
> ```
> $tcpClient = New-Object System.Net.Sockets.TCPClient
> $tcpClient.Connect("localhost",3240)
> $tcpClient.GetStream().WriteByte(1)
> $tcpClient.Close()
> ```
> 
> That should produce the new 'ProtocolViolationException: client disconnect in the middle of a message' error.

This has the same effect. Even when I only make the connection and don't do the WriteByte part
![image](https://user-images.githubusercontent.com/37407247/185730304-985772bc-b3b9-4b60-830c-274c76160d1a.png)

**Comment by dorssel** [comment]
Alright. I got myself a Server 2012R2 installation. And the good news is: I can reproduce the error! That's the best thing ever. I'll get back to you when it is fixed. Thanks for patiently executing all the tests! It has been very helpful.

**Comment by dorssel** [comment]
Found a (the?) bug!
New PR #434. Installer is at https://github.com/dorssel/usbipd-win/actions/runs/2894253188.
Again: uninstall first, as this uses the same development version 2.3.1.

I have good hopes for this one...

**Comment by BjornHoorelbeke** [comment]
Woop woop!!!
![image](https://user-images.githubusercontent.com/37407247/185746334-e6e8f054-658d-4b89-82d1-baddf96fa280.png)

But...
![image](https://user-images.githubusercontent.com/37407247/185746371-d684f71e-a215-4fa0-987d-7aadeec4fd70.png)

Also, trying to attach without the -d option crashes the server/host into a reboot

**Comment by dorssel** [comment]
Nice, step 1 complete: `usbip list` works.

The next problem (really unrelated!) dooms up. You really need to attach with `-b` alone. It gets pretty far, since the host "reacts" to it. Agreed, the "reaction" is reboot, which is not correct, but nevertheless. It means the client asks for the correct device, and the server recognizes that.

I'll tell you what the server is doing during `attach`. It asks the VBoxUSBMon driver to "fake" the device, i.e. it pretends the device is unplugged, then during replugging it mocks the device ID (so it appears to Windows as if an entirely different device is plugged in now). The mocked device ID triggers the loading of the VBoxUSB driver, which will handle the proxying of all USB requests. So ... drivers are involved (and some weird virtual unplugging/replugging), which is always tricky. The drivers are not mine, they are from Oracle VirtualBox. The good news is, 2012R2 *is* supported by VirtualBox. The bad news is, I cannot fix them if they are broken.

First, try to see if there is anything in Windows EventLog from right before the crash. Maybe some log messages survived.

**Comment by BjornHoorelbeke** [comment]
Strange, but now the host doesn't crash anymore using only the `-r` and `-b` options
This is the error I get:
![image](https://user-images.githubusercontent.com/37407247/185848240-5cbc6ccd-374c-4ac7-8ff1-b127cb54b20e.png)

The eventlog shows this entry twice
```
<Event xmlns="http://schemas.microsoft.com/win/2004/08/events/event">
<System>
  <Provider Name="usbipd-win" /> 
  <EventID Qualifiers="0">2</EventID> 
  <Level>4</Level> 
  <Task>0</Task> 
  <Keywords>0x80000000000000</Keywords> 
  <TimeCreated SystemTime="2022-08-22T05:46:23.000000000Z" /> 
  <EventRecordID>1662592</EventRecordID> 
  <Channel>Application</Channel> 
  <Computer>HOME-SERVER.HOME.local</Computer> 
  <Security /> 
</System>
<EventData>
  <Data>Category: Usbipd.ConnectedClient EventId: 2 Client 192.168.1.92 released device at 1-2 (USB\VID_1A86&PID_7523\5&2CDE0EA6&0&2).</Data> 
</EventData>
</Event>
```

**Comment by dorssel** [comment]
That's an easy one. You need to do
```
sudo modprobe vhci-hcd
```
to load the USBIP module in the kernel.

**Comment by BjornHoorelbeke** [comment]
When I try this I get the following error
![image](https://user-images.githubusercontent.com/37407247/185931774-c17338c2-353e-4442-910b-5ff73b60cbcc.png)

Perhaps I should wait for a version where this has been included?
https://github.com/home-assistant/operating-system/pull/2047

**Comment by dorssel** [comment]
That's sad.
We got this far! ...  Only to find out you'll have to wait for some external dependency. But yes, you either need USBIP statically built in the kernel (which it isn't for this kernel), or as a loadable module (which is disabled on this kernel). On the other hand, the feature request is spot on. Once that is implemented you should be good to go.

Summary:
- The drivers didn't work on Server 2012 ... we fixed that
- TCP keepalives didn't work on Server 2012 ... we fixed that
- Host crashed ... that miraculously fixed itself
- Now we need to wait on actual USBIP support in your kernel.

Let me know how that goes once it is available. Good luck! And thanks for testing and hanging in there. `usbipd-win` is a little better now than it was before.

**Comment by BjornHoorelbeke** [comment]
Hupefully they don't wait long before releasing this.
Thank you for your patience and brains!
